### DISCUSSION TITLE
Database Error: remaining connection slots are reserved for non-replication superuser connections

### DISCUSSION THREAD

**Discussion by canumula** [supabase/supabase#13995] [Troubleshooting]
_This is a copy of a troubleshooting article on Supabase's docs site. It may be missing some details from the original. View the [original article](https://supabase.com/docs/guides/troubleshooting/database-error-remaining-connection-slots-are-reserved-for-non-replication-superuser-connections-3V3nIb)._

This error usually occurs when the database reaches the maximum number of connections allowed based on the compute add-on.

To overcome this, the connections need to be optimized as mentioned here: <https://supabase.com/docs/guides/platform/performance#optimizing-the-number-of-connections>

Additionally, you can try using the connection pool to help solve this issue:
<https://supabase.com/docs/guides/database/connecting-to-postgres/pooling-and-limits#how-connection-pooling-works>

If you're already using connection pooling and still hitting the maximum connections, then it is suggested to upgrade your compute add-on that allows more connections: <https://supabase.com/docs/guides/platform/compute-and-disk>

**Comment by TheOtherBrian1** [comment]
You can inspect your database’s connections with Supabase's Grafana Dashboard. 
There is a [GitHub repo](https://github.com/supabase/supabase-grafana) that outlines the process for setting one up locally or on [Fly.io](http://fly.io/) for free.
